### PR TITLE
Refactor GitHub client usage for merges

### LINKED ISSUE #100: Improve how pull.GithubContext uses PR information
`pull.GithubContext` is a bit weird about how it uses `github.PullRequest` objects:

1. It takes one in the constructor and several functions assume it is non-nil, but the constructor also takes 3 parameters that duplicate information from the PR
2. `Body()` just returns the body value of the saved PR object, but takes a context and returns an error
3. `Title()` just returns the title value of the saved PR object, but takes a context and returns an error

Assuming all callers have a full (or full-enough) `github.PullRequest`, we should only require that as an argument and then simplify any methods that only use data from the PR and make no additional API calls.

### PR DESCRIPTION
**Pull request by bluekeyes**
Add additional methods to `pull.Context` (and clean up some of the existing ones) so that `bulldozer.MergePR` can operate without any direct references to a `github.Client`. This improves testability (although I don't add any new tests here) and enables a fix for #98 that I'll add in a followup PR.

Fixes #100.

### CODE REVIEW

File: bulldozer/merge.go [hunk 1]
```diff
@@ -89,78 +117,75 @@ func MergePR(ctx context.Context, pullCtx pull.Context, client *github.Client, m
...
 				if mergeConfig.DeleteAfterMerge {
-					ref := fmt.Sprintf("refs/heads/%s", pr.Head.GetRef())
+					ref := fmt.Sprintf("refs/heads/%s", head)
 
 					// check other open PRs to make sure that nothing is trying to merge into the ref we're about to delete
-					prs, err := pull.ListOpenPullRequestsForRef(ctx, client, pullCtx.Owner(), pullCtx.Repo(), ref)
+					isTargeted, err := pullCtx.IsTargeted(ctx)
 					if err != nil {
-						logger.Error().Err(errors.WithStack(err)).Msgf("Unable to list open prs against ref %s to compare delete request", ref)
+						logger.Error().Err(err).Msgf("Unable to determine if ref %s is targeted by other open pull requests before deletion", ref)
```
Comment (asvoboda) [inline bulldozer/merge.go:179]: nit: Why the change away from `errors.WithStack(err)`?

File: pull/github_context.go [hunk 2]
```diff
@@ -67,12 +67,24 @@ func (ghc *GithubContext) Locator() string {
 	return fmt.Sprintf("%s/%s#%d", ghc.owner, ghc.repo, ghc.number)
 }
 
-func (ghc *GithubContext) Title(ctx context.Context) (string, error) {
-	return ghc.pr.GetTitle(), nil
```
Comment (asvoboda) [inline pull/github_context.go:71]: 🥇 Huh. How did this pass in CR previously..

File: bulldozer/merge.go [hunk 3]
```diff
@@ -89,78 +117,75 @@ func MergePR(ctx context.Context, pullCtx pull.Context, client *github.Client, m
...
 				}
 			}
 
-			logger.Info().Msgf("Successfully merged pull request for sha %s with message %q", result.GetSHA(), result.GetMessage())
+			logger.Info().Msgf("Successfully merged pull request as SHA %s", sha)
 
-			// Delete ref if owner of BASE and HEAD match
-			// otherwise, its from a fork that we cannot delete
-			if pr.GetBase().GetUser().GetLogin() == pr.GetHead().GetUser().GetLogin() {
+			// if head is qualified (contains ":"), PR is from a fork and we don't have delete permission
```
Comment (asvoboda) [inline bulldozer/merge.go:171]: I think we should address this in a later PR, but if the remote fork has bulldozer installed this assumption is no longer correct. I think it was originally here as a workaround from when this operated using oauth tokens. 

We might be able to attempt a delete action, and just fail if it doesn't work. Not blocking, though.

File: bulldozer/merge.go [hunk 4]
```diff
@@ -181,12 +206,7 @@ func calculateCommitMessage(ctx context.Context, pullCtx pull.Context, option Sq
 	commitMessage := ""
 	switch option.Body {
 	case PullRequestBody:
-		body, err := pullCtx.Body(ctx)
-		if err != nil {
-			return "", err
-		}
-
-		commitMessage = body
+		commitMessage = pullCtx.Body()
```
Comment (asvoboda) [inline bulldozer/merge.go:214]: 🎉

**Review by asvoboda — APPROVED** [review]
+1 to the additional cleanup, as well.

File: bulldozer/merge.go [hunk 5]
```diff
@@ -29,19 +29,47 @@ import (
 	"github.com/palantir/bulldozer/pull"
 )
 
-const MaxPullRequestPollCount = 5
+type Merger interface {
+	// Merge merges the pull request in the context using the commit message
+	// and options. It returns the SHA of the merge commit on success.
+	Merge(ctx context.Context, pullCtx pull.Context, message string, options *github.PullRequestOptions) (string, error)
```
Comment (jmcampanini) [inline bulldozer/merge.go]: non-blocking: is it worth also abstracting out the `PullRequestOptions`, so Merger can have no ties to the `github` package? it feels like busy work to do that, but wanted to ask

File: bulldozer/merge.go [hunk 6]
```diff
@@ -89,78 +117,75 @@ func MergePR(ctx context.Context, pullCtx pull.Context, client *github.Client, m
...
-					logger.Info().Msgf("Merge rejected due to unsatisfied condition %q", gerr.Message)
+					logger.Info().Msgf("Merge rejected due to unsatisfied condition: %q", gerr.Message)
 					return
 				case http.StatusConflict:
-					logger.Info().Msgf("Merge rejected due to being invalid %q", gerr.Message)
+					logger.Info().Msgf("Merge rejected due to being invalid: %q", gerr.Message)
 					return
 				default:
-					logger.Error().Err(errors.WithStack(err)).Msgf("Merge failed unexpectedly %q", gerr.Message)
+					logger.Error().Msgf("Merge failed with unexpected status: %d: %q", gerr.Response.StatusCode, gerr.Message)
```
Comment (jmcampanini) [inline bulldozer/merge.go:164]: losing the stacktrace on purpose?

**Review by jmcampanini — APPROVED** [review]

File: bulldozer/merge.go (hunk 5 above)
Comment (bluekeyes) [reply bulldozer/merge.go]: I think that makes sense, mostly because it would remove the split between the commit message and the other options.

File: bulldozer/merge.go (hunk 6 above)
Comment (bluekeyes) [reply bulldozer/merge.go:164]: Yeah, we already know that the error in this case is an HTTP failure, so the important part is the status code and the message. It's also unambiguous where the error is generated.

File: bulldozer/merge.go (hunk 3 above)
Comment (bluekeyes) [reply bulldozer/merge.go:171]: If we want to support this, I consider it a new feature that requires more refactoring: even if the fork has Bulldozer installed, we need to look up the installation ID and construct a new client using that ID.

I'm also not sure how useful this would be. It seems odd to need Bulldozer for merges on a repository that mostly submits PRs to an upstream repo, so would the only reason to install it be so that it could delete branches?

File: bulldozer/merge.go (hunk 1 above)
Comment (bluekeyes) [reply bulldozer/merge.go:179]: I assume that `pull.Context` implementations already wrap errors as needed, so adding a stack again is redundant. It was useful before because the code was mostly dealing with the GitHub client directly.
